File: render-resources/src/com/android/tools/configurations/DeviceStateResolver.java
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.configurations;

import static com.android.tools.configurations.ConfigurationListener.CFG_DEVICE;
import static com.android.tools.configurations.ConfigurationListener.CFG_DEVICE_STATE;

import com.android.ide.common.resources.configuration.ScreenSizeQualifier;
import com.android.annotations.concurrency.Slow;
import com.android.ide.common.resources.configuration.DeviceConfigHelper;
import com.android.ide.common.resources.configuration.FolderConfiguration;
import com.android.ide.common.resources.configuration.ResourceQualifier;
import com.android.ide.common.resources.configuration.ScreenOrientationQualifier;
import com.android.resources.ScreenOrientation;
import com.android.resources.ScreenSize;
import com.android.sdklib.devices.Device;
import com.android.sdklib.devices.State;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Manages the specific hardware constraints and orientation matching logic for a Configuration.
 */
public class DeviceStateResolver {

  public interface Context {
    @NotNull ConfigurationSettings getSettings();

    @NotNull FolderConfiguration getEditedConfig();

    void updateDeviceOverlay();

    @Nullable Device computeBestDevice();
  }

  private final Context myContext;

  @Nullable private Device mySpecificDevice;
  @Nullable private Device myDevice;
  @Nullable private State myState;
  @Nullable private String myStateName;

  public DeviceStateResolver(@NotNull Context context) {
    this.myContext = context;
  }

  public void initFromEditedConfig() {
    ScreenOrientationQualifier qualifier = myContext.getEditedConfig().getScreenOrientationQualifier();
    if (qualifier != null) {
      ScreenOrientation orientation = qualifier.getValue();
      if (orientation != null) {
        myStateName = orientation.getShortDisplayValue();
      }
    }
  }

  @Nullable
  public Device getCachedDevice() { return myDevice; }

  @Nullable
  public State getCachedState() { return myState; }

  @Nullable
  public String getStateName() { return myStateName; }

  @Slow
  @Nullable
  public Device getDevice() {
    if (myDevice != null) return myDevice;

    // This preserves the protected computeBestDevice() extension point for subclasses!
    myDevice = mySpecificDevice != null ? mySpecificDevice : myContext.computeBestDevice();

    myContext.updateDeviceOverlay();
    return myDevice;
  }

  @Nullable
  public State getDeviceState() {
    if (myState == null) {
      Device device = getDevice();
      myState = DeviceState.getDeviceState(device, myStateName);
    }
    return myState;
  }

  public int setDevice(Device device, boolean preserveState) {
    if (mySpecificDevice == device) {
      // The specific device is already set to the correct device so simply clear myDevice
      // which will be re-calculated to be the same as myDevice on the next query.
      myDevice = null;
      return 0;
    }

    Device prevDevice = mySpecificDevice;
    State prevState = myState;

    myDevice = mySpecificDevice = device;
    myContext.updateDeviceOverlay();

    int updateFlags = CFG_DEVICE;

    if (device != null) {
      State state = null;
      // Attempt to preserve the device state?
      if (preserveState && prevDevice != null) {
        if (prevState != null) {
          FolderConfiguration oldConfig = DeviceConfigHelper.getFolderConfig(prevState);
          if (oldConfig != null) {
            String stateName = getClosestMatch(oldConfig, device.getAllStates());
            state = device.getState(stateName);
          }
          else {
            state = device.getState(prevState.getName());
          }
        }
      }
      else if (preserveState && myStateName != null) {
        state = device.getState(myStateName);
      }
      if (state == null) {
        state = device.getDefaultState();
      }
      if (myState != state) {
        updateFlags |= setDeviceStateName(state.getName());
        myState = state;
        updateFlags |= CFG_DEVICE_STATE;
      }
    }
    return updateFlags;
  }

  public int setDeviceState(State state) {
    if (myState != state) {
      int flags = CFG_DEVICE_STATE;
      if (state != null) {
        flags |= setDeviceStateName(state.getName());
      }
      else {
        myStateName = null;
      }
      myState = state;
      return flags;
    }
    return 0;
  }

  public int setDeviceStateName(@Nullable String stateName) {
    ScreenOrientationQualifier qualifier = myContext.getEditedConfig().getScreenOrientationQualifier();
    if (qualifier != null) {
      ScreenOrientation orientation = qualifier.getValue();
      if (orientation != null) {
        stateName = orientation.getShortDisplayValue(); // Also used as state names
      }
    }

    if (!Objects.equals(stateName, myStateName)) {
      myStateName = stateName;
      myState = null;
      return CFG_DEVICE_STATE;
    }
    return 0;
  }

  public int setEffectiveDevice(@Nullable Device device, @Nullable State state) {
    int updateFlags = 0;
    if (myDevice != device) {
      updateFlags = CFG_DEVICE;
      myDevice = device;
      myContext.updateDeviceOverlay();
    }
    if (myState != state) {
      myState = state;
      myStateName = state != null ? state.getName() : null;
      updateFlags |= CFG_DEVICE_STATE;
    }
    return updateFlags;
  }

  @Nullable
  public ScreenSize getScreenSize() {
    State deviceState = getDeviceState();
    if (deviceState != null) {
      FolderConfiguration folderConfig = DeviceConfigHelper.getFolderConfig(deviceState);
      if (folderConfig != null) {
        ScreenSizeQualifier qualifier = folderConfig.getScreenSizeQualifier();
        assert qualifier != null;
        return qualifier.getValue();
      }
    }

    Device device = getDevice();
    if (device != null) {
      for (State state : device.getAllStates()) {
        FolderConfiguration folderConfig = DeviceConfigHelper.getFolderConfig(state);
        if (folderConfig != null) {
          ScreenSizeQualifier qualifier = folderConfig.getScreenSizeQualifier();
          assert qualifier != null;
          return qualifier.getValue();
        }
      }
    }
    return null;
  }

  @Nullable
  public State getNextDeviceState(@Nullable State from) {
    Device device = getDevice();
    if (device == null) return null;

    List<State> states = device.getAllStates();
    for (int i = 0; i < states.size(); i++) {
      if (states.get(i) == from) {
        return states.get((i + 1) % states.size());
      }
    }

    if (from != null) {
      String name = from.getName();
      for (int i = 0; i < states.size(); i++) {
        if (states.get(i).getName().equals(name)) {
          return states.get((i + 1) % states.size());
        }
      }
    }
    return null;
  }

  @Nullable
  private static String getClosestMatch(@NotNull FolderConfiguration oldConfig, @NotNull List<State> states) {
    List<State> list1 = new ArrayList<>(states.size());
    List<State> list2 = new ArrayList<>(states.size());

    list1.addAll(states);

    final int count = FolderConfiguration.getQualifierCount();
    for (int i = 0; i < count; i++) {
      for (State s : list1) {
        ResourceQualifier oldQualifier = oldConfig.getQualifier(i);
        FolderConfiguration folderConfig = DeviceConfigHelper.getFolderConfig(s);
        ResourceQualifier newQualifier = folderConfig != null ? folderConfig.getQualifier(i) : null;

        if (oldQualifier == null) {
          if (newQualifier == null) {
            list2.add(s);
          }
        }
        else if (oldQualifier.equals(newQualifier)) {
          list2.add(s);
        }
      }
      if (list2.size() == 1) {
        return list2.get(0).getName();
      }
      if (!list2.isEmpty()) {
        list1.clear();
        list1.addAll(list2);
        list2.clear();
      }
    }
    if (!list1.isEmpty()) {
      return list1.get(0).getName();
    }
    return null;
  }

  public void copyFrom(DeviceStateResolver other) {
    this.mySpecificDevice = other.mySpecificDevice;
    this.myDevice = other.myDevice;
    this.myState = other.myState;
    this.myStateName = other.myStateName;
  }
}

File: render-resources/src/com/android/tools/configurations/DeviceStateResolver.kt
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.configurations

import com.android.annotations.concurrency.Slow
import com.android.ide.common.resources.configuration.DeviceConfigHelper
import com.android.ide.common.resources.configuration.FolderConfiguration
import com.android.resources.ScreenSize
import com.android.sdklib.devices.Device
import com.android.sdklib.devices.State
import com.android.tools.configurations.ConfigurationListener.CFG_DEVICE
import com.android.tools.configurations.ConfigurationListener.CFG_DEVICE_STATE
import java.util.ArrayList

/** Manages the specific hardware constraints and orientation matching logic for a Configuration. */
class DeviceStateResolver(private val context: Context) {

  interface Context {
    val settings: ConfigurationSettings
    val editedConfig: FolderConfiguration

    fun updateDeviceOverlay()

    fun computeBestDevice(): Device?
  }

  private var specificDevice: Device? = null
  private var device: Device? = null
  private var state: State? = null

  var stateName: String? = null
    private set

  val cachedDevice: Device?
    get() = device

  val cachedState: State?
    get() = state

  fun initFromEditedConfig() {
    val qualifier = context.editedConfig.screenOrientationQualifier
    stateName = qualifier?.value?.shortDisplayValue ?: stateName
  }

  @Slow
  fun getDevice(): Device? {
    if (device != null) return device

    // This preserves the protected computeBestDevice() extension point for subclasses!
    device = specificDevice ?: context.computeBestDevice()
    context.updateDeviceOverlay()
    return device
  }

  fun getDeviceState(): State? {
    if (state == null) {
      val currentDevice = getDevice()
      state = currentDevice.getDeviceState(stateName)
    }
    return state
  }

  fun setDevice(newDevice: Device?, preserveState: Boolean): Int {
    if (specificDevice === newDevice) {
      // Even if the specific device hasn't changed, make sure the cached device is null
      // to force recomputation of the "best" device if necessary.
      device = null
      return 0
    }

    val prevDevice = specificDevice
    val prevState = state

    specificDevice = newDevice
    device = newDevice
    context.updateDeviceOverlay()

    var updateFlags = CFG_DEVICE

    if (newDevice != null) {
      var newState: State? = null
      // Attempt to preserve the device state?
      if (preserveState && prevDevice != null) {
        if (prevState != null) {
          val oldConfig = DeviceConfigHelper.getFolderConfig(prevState)
          if (oldConfig != null) {
            val matchName = getClosestMatch(oldConfig, newDevice.allStates)
            newState = newDevice.getState(matchName)
          } else {
            newState = newDevice.getState(prevState.name)
          }
        }
      } else if (preserveState && stateName != null) {
        newState = newDevice.getState(stateName)
      }

      if (newState == null) {
        newState = newDevice.defaultState
      }

      if (state !== newState) {
        updateFlags = updateFlags or setDeviceStateName(newState?.name)
        state = newState
        updateFlags = updateFlags or CFG_DEVICE_STATE
      }
    }
    return updateFlags
  }

  fun setDeviceState(newState: State?): Int {
    if (state !== newState) {
      var flags = CFG_DEVICE_STATE
      if (newState != null) {
        flags = flags or setDeviceStateName(newState.name)
      } else {
        stateName = null
      }
      state = newState
      return flags
    }
    return 0
  }

  fun setDeviceStateName(newStateName: String?): Int {
    var actualStateName = newStateName
    val qualifier = context.editedConfig.screenOrientationQualifier
    if (qualifier != null) {
      val orientation = qualifier.value
      if (orientation != null) {
        actualStateName = orientation.shortDisplayValue // Also used as state names
      }
    }

    if (actualStateName != stateName) {
      stateName = actualStateName
      state = null
      return CFG_DEVICE_STATE
    }
    return 0
  }

  fun setEffectiveDevice(newDevice: Device?, newState: State?): Int {
    var updateFlags = 0
    if (device !== newDevice) {
      updateFlags = CFG_DEVICE
      device = newDevice
      context.updateDeviceOverlay()
    }

    if (state !== newState) {
      state = newState
      stateName = newState?.name
      updateFlags = updateFlags or CFG_DEVICE_STATE
    }
    return updateFlags
  }

  fun getScreenSize(): ScreenSize? {
    val deviceState = getDeviceState()
    if (deviceState != null) {
      val folderConfig = DeviceConfigHelper.getFolderConfig(deviceState)
      if (folderConfig != null) {
        val qualifier = folderConfig.screenSizeQualifier
        assert(qualifier != null)
        return qualifier?.value
      }
    }

    val currentDevice = getDevice()
    if (currentDevice != null) {
      for (s in currentDevice.allStates) {
        val folderConfig = DeviceConfigHelper.getFolderConfig(s)
        if (folderConfig != null) {
          val qualifier = folderConfig.screenSizeQualifier
          assert(qualifier != null)
          return qualifier?.value
        }
      }
    }
    return null
  }

  fun getNextDeviceState(from: State?): State? {
    val currentDevice = getDevice() ?: return null
    val states = currentDevice.allStates

    if (states.isEmpty()) return null

    for (i in states.indices) {
      if (states[i] === from) {
        return states[(i + 1) % states.size]
      }
    }

    if (from != null) {
      val name = from.name
      for (i in states.indices) {
        if (states[i].name == name) {
          return states[(i + 1) % states.size]
        }
      }
    }
    return null
  }

  fun copyFrom(other: DeviceStateResolver) {
    this.specificDevice = other.specificDevice
    this.device = other.device
    this.state = other.state
    this.stateName = other.stateName
  }

  companion object {
    private fun getClosestMatch(oldConfig: FolderConfiguration, states: List<State>): String? {
      val list1 = ArrayList(states)
      val list2 = ArrayList<State>(states.size)

      val count = FolderConfiguration.getQualifierCount()
      for (i in 0 until count) {
        for (s in list1) {
          val oldQualifier = oldConfig.getQualifier(i)
          val folderConfig = DeviceConfigHelper.getFolderConfig(s)
          val newQualifier = folderConfig?.getQualifier(i)

          if (oldQualifier == null) {
            if (newQualifier == null) {
              list2.add(s)
            }
          } else if (oldQualifier == newQualifier) {
            list2.add(s)
          }
        }

        if (list2.size == 1) {
          return list2[0].name
        }

        if (list2.isNotEmpty()) {
          list1.clear()
          list1.addAll(list2)
          list2.clear()
        }
      }

      if (list1.isNotEmpty()) {
        return list1[0].name
      }

      return null
    }
  }
}
